flutter/lib/src/frames_tracking/sentry_delayed_frames_tracker.dart

```diff
@@ -0,0 +1,303 @@
+// ignore_for_file: invalid_use_of_internal_member
+
+import 'dart:math';
+
+import 'package:meta/meta.dart';
+import '../../sentry_flutter.dart';
+
+/// 30s span duration at 120fps = 3600 frames
+/// This is just an upper limit, ensuring that the buffer does not grow
+/// indefinitely in case of a long running span.
+/// Realistically this won't happen since we only track slow or frozen frames
+/// but it's going to help us safeguard in the rare cases when it happens.
+///
+/// If this limit is reached, we stop collecting frames until all active spans
+/// have finished processing.
+const maxFramesCount = 3600;
+
+/// The duration at which we consider a frame 'frozen'
+const _frozenFrameThreshold = Duration(milliseconds: 700);
+
+/// Singleton frame tracker to collect delayed frames processed by the Flutter SDK.
+///
+/// The tracker needs to be initialized first via [SentryWidgetsBindingMixin.initializesFrameTracker]
+/// otherwise the tracker won't collect frames.
+///
+/// The order in which [startFrame] and [endFrame] is called is sequential
+/// and depends on Flutter to be accurate and precise. Each tracked frame is
+/// aimed to replicate the build duration that you would receive from [addTimingsCallback].
+@internal
+class SentryDelayedFramesTracker {
+  SentryDelayedFramesTracker(this._options, this._expectedFrameDuration);
+
+  /// List of frame timings that holds delayed frames (slow and frozen frames).
+  /// We don't keep track of normal frames since we can estimate the number of
+  /// normal frames based on the span duration and the expected frame duration.
+  /// Since startFrame and endFrame is always called sequentially by Flutter we
+  /// don't need a SplayTree
+  final List<SentryFrameTiming> _delayedFrames = [];
+  final SentryFlutterOptions _options;
+  final Duration _expectedFrameDuration;
+  DateTime? _currentFrameStartTimestamp;
+  bool _isTrackingActive = false;
+
+  /// Marks the start of a frame.
+  @pragma('vm:prefer-inline')
+  void startFrame() {
+    if (!_isTrackingActive || _options.enableFramesTracking == false) {
+      return;
+    }
+
+    _currentFrameStartTimestamp = _options.clock();
+  }
+
+  /// Marks the end of a frame.
+  @pragma('vm:prefer-inline')
+  void endFrame() {
+    if (!_isTrackingActive || !_options.enableFramesTracking) {
+      return;
+    }
+
+    final endTimestamp = _options.clock();
+    final startTimestamp = _currentFrameStartTimestamp;
+
+    if (!_isFrameValid(startTimestamp, endTimestamp)) {
+      _resetCurrentFrame();
+      return;
+    }
+
+    _processFrame(startTimestamp!, endTimestamp);
+  }
+
+  bool _isFrameValid(DateTime? startTimestamp, DateTime? endTimestamp) {
+    return startTimestamp != null && endTimestamp != null;
```

Comment:
Should we also check of startTimestamp < endTimestamp here?

Reply:
Also, endTimestamp from options.clock() will always be non-null. No need to have an optional param here and do the null check.

Reply:
> Should we also check of startTimestamp < endTimestamp here?

we should, yeah